homeassistant/components/idasen_desk/switch.py

```diff
@@ -0,0 +1,93 @@
+"""Idasen Desk switch to turn bluetooth connection on/off."""
+from __future__ import annotations
+
+import logging
+from typing import Any
+
+from homeassistant.components.switch import SwitchDeviceClass, SwitchEntity
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import ATTR_NAME, STATE_OFF
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers.device_registry import DeviceInfo
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.restore_state import RestoreEntity
+from homeassistant.helpers.update_coordinator import CoordinatorEntity
+
+from . import DeskData, IdasenDeskCoordinator
+from .const import DOMAIN
+
+RECONNECT_RETRY_SECONDS = 30
+
+# Initialize the logger
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the cover platform for Idasen Desk."""
+    data: DeskData = hass.data[DOMAIN][entry.entry_id]
+    async_add_entities(
+        [IdasenDeskSwitch(data.address, data.device_info, data.coordinator)]
+    )
+
+
+class IdasenDeskSwitch(CoordinatorEntity, SwitchEntity, RestoreEntity):
+    """Idasen Desk connection switch."""
+
+    _attr_device_class = SwitchDeviceClass.SWITCH
+
+    def __init__(
+        self,
+        address: str,
+        device_info: DeviceInfo,
+        coordinator: IdasenDeskCoordinator,
+    ) -> None:
+        """Initialize the connection switch."""
+        super().__init__(coordinator)
+        self._address = address
+        self._attr_name = f"{device_info[ATTR_NAME]} Connection"
+        self._attr_unique_id = address
+        self._attr_device_info = device_info
+        self._attr_is_on = False
+
+    async def async_added_to_hass(self) -> None:
+        """Run when entity about to be added."""
+        await super().async_added_to_hass()
+        last_state = await self.async_get_last_state()
+        if last_state and last_state.state == STATE_OFF:
+            await self.async_turn_off()
+        else:
+            # If last state is None it should also turn on, since it should be
+            # on by default unless the user has explicitly turned it off.
+            await self.async_turn_on()
```

Comment:
After looking through this, I realized this is a virtual entity as it doesn't remember the state of the device, but instead represents the state of Home Assistant configuration.

We currently don't allow virtual entities to be added to integrations.

Once option would be to make it a config entry option.

Reply:
The reasoning for the switch is the limited number of active connections on bluetooth radios (ESPHome only allows 3, for example). If you have multiple devices (not only desks), you can have an automation to only connect to the desk when you want to move it.

Reply:
An alternative would be too add `homeassistant.disable_config_entry` and `homeassistant.enable_config_entry`, as we already have `homeassistant.reload_config_entry`

Reply:
Looks like there is a feature request for that https://community.home-assistant.io/t/service-to-disable-and-enable-integrations/576101/15

Reply:
It looks like a previous PR has was not accepted that implemented that https://github.com/home-assistant/core/pull/77457

Reply:
In in ideal world the desk broadcasts it state via the advertisement and you only connect to it when you need to make a change.

I don't have one of these so I have no idea whats in the advertisement

Reply:
Another option would be to add a button entity to connect the desk and one to disconnect the desk.  That would avoid the whole virtual entity problem.

Reply:
Unfortunately the desk only sends data to the device connected to it. Also, it only connects to a single device at a time, so the user needs to be able to disconnect it from HA to use the Linak app.

Is the only reason this is not allowed the fact that the state of the switch is virtual? That is why buttons would be accepted instead?

Reply:
> Is the only reason this is not allowed the fact that the state of the switch does not reflect some state on the device? That is why buttons would be accepted instead?

That is exactly the reason.

The test for a virtual entity is generally this:

Delete the integration/config entry, add it back, if the entity is not in the same state as it was before, it might be considered a virtual entity.

Reply:
Right. I'm fine in switching to buttons. I will update this PR.

I do like the fact that the switch gave some feedback about whether or not the integration was trying to connect or not (instead of just connected/disconnected), but it is not critical I guess.

Reply:
Another downside with the buttons is that after HA starts, it will always connect to the desk, while the switch restored the last state.

Reply:
You could also an options flow to change the connected state at startup

Reply:
Good idea